### PR TITLE
DOC: Add new pages showing RunEngine API docs

### PR DESCRIPTION
**Pull request by danielballan**

### CODE REVIEW

**Comment by tacaswell** [comment]
Sorry, I broke this by fixing .travis in #682

**Comment by danielballan** [comment]
@tacaswell Your point (offline I think, or in Slack) about rethinking the Reader and Mover API -- and also reusing more or ophyd -- is well taken. Is it worth documenting the current situation a little better anyway? I would vote "yes" unless we are confident we'll commit to a new API before the next cycle.

**Comment by tacaswell** [comment]
I am against documenting the current example classes better until we sort out what we want to do with them.  End users should not be using them (other than the few instances we provide) and they seem needlessly complicated for what they do.

**Comment by danielballan** [comment]
FWIW I don't feel so negatively about them and I would like to see users developing more interesting simulated hardware. Let's sort out a permanent direction for these soon.

**Comment by danielballan** [comment]
> I am against documenting the current example classes better until we sort out what we want to do with them. End users should not be using them (other than the few instances we provide) and they seem needlessly complicated for what they do.

The current example classes are still not documented. Do you still want to hold off on this, @tacaswell?

**Comment by CJ-Wright** [comment]
What is the status of the simulated `example` classes? We are building some infrastructure with them so it would be nice to know if they are going to still be around and/or what their replacements will be.

**Comment by danielballan** [comment]
I can't envision how we could make Reader or Mover any simpler than they are now, or what improvements would justify an API break. But I'm not sure if Tom's concerns relate to those or the more specialized one-offs like SynGauss.

Get Outlook for iOS<https://aka.ms/o0ukef>



On Mon, Jul 10, 2017 at 6:00 PM -0400, "Christopher J. Wright" <notifications@github.com<mailto:notifications@github.com>> wrote:


What is the status of the simulated example classes? We are building some infrastructure with them so it would be nice to know if they are going to still be around and/or what their replacements will be.

**Comment by CJ-Wright** [comment]
I'm mostly worried about `ReaderWithFilestore`, although `Mover` and `Reader` are critical too.

**Comment by tacaswell** [comment]
I am still :-1: on documenting the examples and to @CJ-Wright building anything on top of them.

Ophyd now imports with out epics, you should use `Device` and `Signal` to build synthetic devices.

**Comment by danielballan** [comment]
When one sits down the actually write these for testing or simulation, I think the sparseness of Reader(name, {field: func}) is very appealing. Perhaps we can rebuild that API on top of Device. Also remember that ophyd is not available in analysis envs, as a layer of security.

**Comment by danielballan** [comment]
I removed the commits related to `bluesky.examples`, punting that discussion to a separate issue. This PR now only contains RunEngine and Dispatcher API documentation, with associated changes to the docstrings in `run_engine.py`.

File: bluesky/run_engine.py [hunk 1]
```diff
@@ -288,6 +280,52 @@ def __init__(self, md=None, *, loop=None, preprocessors=None,
 
         self.loop.call_soon(self._check_for_signals)
 
+    def subscribe(self, name, func):
```
Comment (tacaswell) [inline bluesky/run_engine.py]: This should be flipped in accoradnce with #691

File: doc/source/run_engine_api.rst [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+.. autoclass:: RunEngine
+
+    .. automethod:: __call__
+
+    The RunEngine maintains a callback registry of functions that receive any
+    :doc:`documents` generated by plan execution. These methods add and remove
+    functions from that registry.
+
+    .. automethod:: subscribe
```
Comment (tacaswell) [inline doc/source/run_engine_api.rst]: I think using `autosummary` here would be better so each method gets it's own page and these are just tables. 

Not a blocking concern.

Comment (danielballan) [inline doc/source/run_engine_api.rst]: I gave it a try (in a currently stashed local commit) and I don't like the results as much. I do like autosummary for something like our table built-in plans, where the functions are fairly self-describing and its docstring's top-line sentence is enough to explain its purpose. But for many RunEngine methods, the purpose requires more explanation. It's nice to put more detail up front.

File: bluesky/run_engine.py (hunk 1 above)
Comment (danielballan) [inline bluesky/run_engine.py]: Updated and force-pushed.

File: doc/source/run_engine_api.rst (hunk 2 above)
Comment (danielballan) [inline doc/source/run_engine_api.rst]: Actually, sleeping on this, maybe tables _are_ better. But it seems that sphinx doesn't give methods separate pages the way it does for functions. Our choices are (1) automethod style or (2) autosummary with no links, and therefore no access to the details like "Examples" or "See also".

Comment (tacaswell) [inline doc/source/run_engine_api.rst]: There is a way to make it do it, see http://matplotlib.org/api/axes_api.html#basic

File: bluesky/run_engine.py [hunk 3]
```diff
@@ -2011,13 +2067,17 @@ def subscribe(self, func, name='all'):
         func: callable
             expecting signature like ``f(name, document)``
             where name is a string and document is a dict
-        name: {'start', 'descriptor', 'event', 'stop', 'all'}, optional
-            Defaults to 'all'
+        name: string
+            one of {'start', 'descriptor', 'event', 'stop', 'all'}
```
Comment (tacaswell) [inline bluesky/run_engine.py]: https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt see section 4 the convention is to list
the valid values (and the default should go first).

File: doc/source/run_engine_api.rst (hunk 2 above)
Comment (danielballan) [inline doc/source/run_engine_api.rst]: OK, I'm up for that, but this feels like the beginning of a three-hour fight with sphinx that I don't have time for before Chicago. Punting to https://github.com/NSLS-II/bluesky/issues/701

File: bluesky/run_engine.py (hunk 3 above)
Comment (danielballan) [inline bluesky/run_engine.py]: Added new commit to address

**Pull request merged**